### PR TITLE
Add deferrable `AzureDataFactoryPipelineRunStatusSensor`

### PR DESCRIPTION
**Pull request by phanikumv**
This PR donates the `AzureDataFactoryPipelineRunStatusAsyncSensor` from [astronomer-providers](https://github.com/astronomer/astronomer-providers) repo to OSS

cc @josh-fell @kaxil 

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: airflow/providers/microsoft/azure/sensors/data_factory.py [hunk 1]
```diff
@@ -78,3 +84,60 @@ def poke(self, context: Context) -> bool:
...
+        **kwargs: Any,
+    ):
+        # TODO: Remove once deprecated
+        if poll_interval:
+            self.poke_interval = poll_interval
+            warnings.warn(
+                "Argument `poll_interval` is deprecated and will be removed "
+                "in a future release.  Please use  `poke_interval` instead.",
+                DeprecationWarning,
+                stacklevel=2,
```
Comment (kaxil) [inline airflow/providers/microsoft/azure/sensors/data_factory.py]: Why deprecate when we have not even mergred, released? Copy/paste error from `astronomer-providers`?

**Review by kaxil — COMMENTED** [review]
Could you add any docs or example DAG for this?

File: airflow/providers/microsoft/azure/hooks/data_factory.py [hunk 2]
```diff
@@ -1039,3 +1048,120 @@ def test_connection(self) -> tuple[bool, str]:
...
+    @wraps(func)
+    async def wrapper(*args: Any, **kwargs: Any) -> Any:
+        bound_args = signature.bind(*args, **kwargs)
+
+        async def bind_argument(arg: Any, default_key: str) -> None:
+            # Check if arg was not included in the function signature or, if it is, the value is not provided.
+            if arg not in bound_args.arguments or bound_args.arguments[arg] is None:
+                self = args[0]
+                conn = await sync_to_async(self.get_connection)(self.conn_id)
+                default_value = conn.extra_dejson.get(default_key)
```
Comment (josh-fell) [inline airflow/providers/microsoft/azure/hooks/data_factory.py]: Can you update this to match the sync version? Prefixing connection extras with `extra__...` is no longer needed, but we do keep the check for backwards compat.

File: airflow/providers/microsoft/azure/hooks/data_factory.py [hunk 3]
```diff
@@ -1039,3 +1048,120 @@ def test_connection(self) -> tuple[bool, str]:
...
+        self.conn_id = azure_data_factory_conn_id
+        super().__init__(azure_data_factory_conn_id=azure_data_factory_conn_id)
+
+    async def get_async_conn(self) -> AsyncDataFactoryManagementClient:
+        """Get async connection and connect to azure data factory"""
+        if self._conn is not None:
+            return self._conn
+
+        conn = await sync_to_async(self.get_connection)(self.conn_id)
+        tenant = conn.extra_dejson.get("extra__azure_data_factory__tenantId")
```
Comment (josh-fell) [inline airflow/providers/microsoft/azure/hooks/data_factory.py]: Same idea here. Should use the `get_field()` function to retrieve the extras.

File: tests/providers/microsoft/azure/hooks/test_azure_data_factory.py [hunk 4]
```diff
@@ -708,3 +723,214 @@ def test_backcompat_prefix_both_prefers_short(mock_connect):
         hook = AzureDataFactoryHook("my_conn")
         hook.delete_factory(factory_name="n/a")
         mock_connect.return_value.factories.delete.assert_called_with("non-prefixed", "n/a")
+
+
+class TestAzureDataFactoryAsyncHook:
+    @pytest.mark.asyncio
+    @pytest.mark.parametrize(
```
Comment (josh-fell) [inline tests/providers/microsoft/azure/hooks/test_azure_data_factory.py]: Why parametrize if each status is its own explicit test? You can inject test case `ids` like `pytest.mark.parametrize(..., ids=["test_status1", "test_status2"])`

**Review by josh-fell — CHANGES_REQUESTED** [review]

File: airflow/providers/microsoft/azure/sensors/data_factory.py (hunk 1 above)
Comment (phanikumv) [reply airflow/providers/microsoft/azure/sensors/data_factory.py]: yes it is a typo. Corrected it.

File: tests/providers/microsoft/azure/hooks/test_azure_data_factory.py (hunk 4 above)
Comment (phanikumv) [reply tests/providers/microsoft/azure/hooks/test_azure_data_factory.py]: yeah, initially I thought of putting all the status in one test by parameterization..but asynctest tests on Python 3.7 wont run like that, and will give the below error

```FAILED test_azure_data_factory.py::TestAzureDataFactoryAsyncHook::test_get_adf_pipeline_run_status_queued[InProgress] - AttributeError: '_patch' object has no attribute '_exit_stack'```

when used in combination with pytest.mark.parametrize it leads to mock __enter()__ being called only once per test case (regardless of number of "subcases" created with parametrisation), but the patch __exit()__ is called after each of the parametrisation "subtests", so only the first "subtest" passes and all other fail. Here is a link about the issue for reference

https://github.com/Martiusweb/asynctest/issues/152#issue-666326338

File: airflow/providers/microsoft/azure/hooks/data_factory.py (hunk 3 above)
Comment (phanikumv) [reply airflow/providers/microsoft/azure/hooks/data_factory.py]: done

File: airflow/providers/microsoft/azure/hooks/data_factory.py (hunk 2 above)
Comment (phanikumv) [reply airflow/providers/microsoft/azure/hooks/data_factory.py]: done

**Comment by phanikumv** [comment]
> Could you add any docs or example DAG for this?

this is done, added docs and example DAG

File: airflow/providers/microsoft/azure/hooks/data_factory.py [hunk 5]
```diff
@@ -1039,3 +1048,124 @@ def test_connection(self) -> tuple[bool, str]:
...
+
+
+class AzureDataFactoryAsyncHook(AzureDataFactoryHook):
+    """
+    An Async Hook that connects to Azure DataFactory to perform pipeline operations
+
+    :param azure_data_factory_conn_id: The :ref:`Azure Data Factory connection id<howto/connection:adf>`.
+    """
+
+    def __init__(self, azure_data_factory_conn_id: str):
```
Comment (josh-fell) [inline airflow/providers/microsoft/azure/hooks/data_factory.py]: WDYT about making `azure_data_factory_conn_id` optional like it is in AzureDataFactoryHook?

File: airflow/providers/microsoft/azure/hooks/data_factory.py [hunk 6]
```diff
@@ -1039,3 +1048,124 @@ def test_connection(self) -> tuple[bool, str]:
...
+        self._async_conn: AsyncDataFactoryManagementClient = None
+        self.conn_id = azure_data_factory_conn_id
+        super().__init__(azure_data_factory_conn_id=azure_data_factory_conn_id)
+
+    async def get_async_conn(self) -> AsyncDataFactoryManagementClient:
+        """Get async connection and connect to azure data factory"""
+        if self._conn is not None:
```
Comment (josh-fell) [inline airflow/providers/microsoft/azure/hooks/data_factory.py]: I think this function should use `self._async_conn` instead?

File: airflow/providers/microsoft/azure/hooks/data_factory.py [hunk 7]
```diff
@@ -1039,3 +1048,124 @@ def test_connection(self) -> tuple[bool, str]:
...
+        resource_group_name: str | None = None,
+        factory_name: str | None = None,
+        **config: Any,
+    ) -> PipelineRun:
+        """
+        Connect to Azure Data Factory asynchronously to get the pipeline run details by run id
+
+        :param run_id: The pipeline run identifier.
+        :param resource_group_name: The resource group name.
+        :param factory_name: The factory name.
```
Comment (josh-fell) [inline airflow/providers/microsoft/azure/hooks/data_factory.py:1147]:
```suggestion
        :param factory_name: The factory name.
        :param config: Extra parameters for the ADF client.
```
To be consistent with the other sync functions and to let users know they can pass in additional configuration should they choose to use this function directly.

File: airflow/providers/microsoft/azure/sensors/data_factory.py [hunk 8]
```diff
@@ -78,3 +83,52 @@ def poke(self, context: Context) -> bool:
...
+
+class AzureDataFactoryPipelineRunStatusAsyncSensor(AzureDataFactoryPipelineRunStatusSensor):
+    """
+    Checks the status of a pipeline run asynchronously.
+
+    :param azure_data_factory_conn_id: The connection identifier for connecting to Azure Data Factory.
+    :param run_id: The pipeline run identifier.
+    :param resource_group_name: The resource group name.
+    :param factory_name: The data factory name.
+    :param poll_interval: polling period in seconds to check for the status
```
Comment (josh-fell) [inline airflow/providers/microsoft/azure/sensors/data_factory.py]:
```suggestion
    :param poke_interval: polling period in seconds to check for the status
```

File: airflow/providers/microsoft/azure/hooks/data_factory.py (hunk 6 above)
Comment (josh-fell) [reply airflow/providers/microsoft/azure/hooks/data_factory.py]: And `self._async_conn` should be returned too so the initial `is not None` check acts as a form of "caching".

File: tests/providers/microsoft/azure/hooks/test_azure_data_factory.py [hunk 9]
```diff
@@ -708,3 +724,245 @@ def test_backcompat_prefix_both_prefers_short(mock_connect):
...
+            {
+                "extra__azure_data_factory__tenantId": "tenantId",
+                "extra__azure_data_factory__resource_group_name": RESOURCE_GROUP_NAME,
+                "extra__azure_data_factory__factory_name": DATAFACTORY_NAME,
+            }
+        ],
+    )
+    @async_mock.patch(f"{MODULE}.hooks.data_factory.AzureDataFactoryAsyncHook.get_connection")
+    async def test_get_async_conn_key_error_tenantId(self, mock_connection, mock_connection_params):
+        """Test get_async_conn function with raising key error"""
```
Comment (josh-fell) [inline tests/providers/microsoft/azure/hooks/test_azure_data_factory.py]: Would you mind updating the docstrings for this function and `test_get_async_conn_key_error_subscriptionId`? They are both the same.

Although I think the function names should be switched? `test_get_async_conn_key_error_tenantId` is testing that  a subscription ID is not part of the connection while `test_get_async_conn_key_error_subscriptionId` is testing that tenant ID is required when using login/password.

**Review by josh-fell — COMMENTED** [review]
Looks pretty good! Just a few comments and questions.

File: tests/providers/microsoft/azure/hooks/test_azure_data_factory.py (hunk 9 above)
Comment (phanikumv) [reply tests/providers/microsoft/azure/hooks/test_azure_data_factory.py]: done

File: airflow/providers/microsoft/azure/sensors/data_factory.py (hunk 8 above)
Comment (phanikumv) [reply airflow/providers/microsoft/azure/sensors/data_factory.py]: fixed

File: airflow/providers/microsoft/azure/hooks/data_factory.py (hunk 7 above)
Comment (phanikumv) [reply airflow/providers/microsoft/azure/hooks/data_factory.py:1147]: done

File: airflow/providers/microsoft/azure/hooks/data_factory.py (hunk 6 above)
Comment (phanikumv) [reply airflow/providers/microsoft/azure/hooks/data_factory.py]: correct , fixed it

File: airflow/providers/microsoft/azure/hooks/data_factory.py (hunk 5 above)
Comment (phanikumv) [reply airflow/providers/microsoft/azure/hooks/data_factory.py]: done

File: airflow/providers/microsoft/azure/hooks/data_factory.py (hunk 6 above)
Comment (josh-fell) [reply airflow/providers/microsoft/azure/hooks/data_factory.py]: `self._async_conn` isn't actually being set anywhere. Can you add that so the `is not None` check here will just return an existing connection object?

Comment (phanikumv) [reply airflow/providers/microsoft/azure/hooks/data_factory.py]: this is done in [264e90d](https://github.com/apache/airflow/pull/29801/commits/264e90dd62f4f6be0498b7df0fb03594b5936153)

**Review by kaxil — APPROVED** [review]

**Review by josh-fell — APPROVED** [review]

**Pull request merged**